### PR TITLE
Issue#19 multilayered opds feeds

### PR DESCRIPTION
**Pull request by kstigen**

### CODE REVIEW

File: src/main/scala/no/gdl/bookapi/controller/OPDSController.scala [hunk 1]
```diff
@@ -8,56 +8,131 @@
...
-
     before() {
       contentType = formats("xml")
     }
 
-    get("/:lang/catalog.atom") {
-      showOpdsForLanguage(params("lang"))
+    get(BookApiProperties.OpdsNavUrl.url) {
+      val navFeeds = feedService.feedsForNavigation(language("lang"))
+      val navUpdated = navFeeds.map(_.updated).sorted.reverse.headOption
```
Comment (runarfu) [inline src/main/scala/no/gdl/bookapi/controller/OPDSController.scala]: `.sorted.reverse` is not 100% intuitive, could this value be named `navLastUpdated` or something similar to more accurately show the intent?

File: src/main/scala/no/gdl/bookapi/controller/OPDSController.scala [hunk 2]
```diff
@@ -8,56 +8,131 @@
...
+    private def navigationFeed(feedUpdated: Option[LocalDate], feeds: => Seq[Feed])(implicit request: HttpServletRequest) = {
+      val lang = language("lang")
+      val selfOpt = feedService.feedForUrl(request.getRequestURI, lang, feedUpdated, Seq(), Seq())
+      selfOpt match {
+        case None =>
+          contentType = "text/plain"
+          NotFound(body = s"No books available for language $lang.")
+
+        case Some(self) => {
+          <feed xmlns="http://www.w3.org/2005/Atom">
```
Comment (runarfu) [inline src/main/scala/no/gdl/bookapi/controller/OPDSController.scala]: Should we have a unit test for the xml generated here and also in `acquisitionFeed`?

File: src/main/scala/no/gdl/bookapi/model/api/Book.scala [hunk 3]
```diff
@@ -98,6 +98,30 @@ case class SearchResult(@(ApiModelProperty@field)(description = "The total numbe
                         @(ApiModelProperty@field)(description = "The chosen language") language: Language,
                         @(ApiModelProperty@field)(description = "The results") results: Seq[Book])
 
+@ApiModel(description = "Information about the editors picks")
```
Comment (runarfu) [inline src/main/scala/no/gdl/bookapi/model/api/Book.scala]: `editor's` (repeated typo in this file)

File: src/main/scala/no/gdl/bookapi/model/api/Book.scala [hunk 4]
```diff
@@ -98,6 +98,30 @@ case class SearchResult(@(ApiModelProperty@field)(description = "The total numbe
...
+                          @(ApiModelProperty@field)(description = "The uuid of the feed") uuid: String)
+
+@ApiModel(description = "Information about an OPDS-Feed")
+case class Feed (feedDefinition: FeedDefinition, title: String, description: Option[String], rel: Option[String], updated: LocalDate, content: Seq[FeedEntry])
+
+@ApiModel(description = "Information about an Entry in an OPDS-Feed")
+case class FeedEntry (@(ApiModelProperty@field)(description = "The book associated with this entry") book: Book,
+                      @(ApiModelProperty@field)(description = "The revision of the feed") categories: Seq[FeedCategory] = Seq())
+
+@ApiModel(description = "Information about an feed category in an opds-feed")
```
Comment (runarfu) [inline src/main/scala/no/gdl/bookapi/model/api/Book.scala]: s/an/a

File: src/main/scala/no/gdl/bookapi/service/FeedService.scala [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    def addJustArrivedCategory(feedEntry: FeedEntry, language: String): FeedEntry = {
+      val title = Messages("new_entries_feed_title")(Lang(language))
+      val url = s"${BookApiProperties.Domain}${justArrivedPath(language)}"
+      feedEntry.copy(
+        categories = feedEntry.categories :+ FeedCategory(url, title, sortOrder = 2))
+    }
+
+    def addLevelCategory(feedEntry: FeedEntry, language: String): FeedEntry = {
+      val level = feedEntry.book.readingLevel.getOrElse(BookApiProperties.DefaultReadingLevel)
+      val readingLevelCategoryTitle = s"${Messages("level_feed_title", level)(Lang(language))}"
```
Comment (runarfu) [inline src/main/scala/no/gdl/bookapi/service/FeedService.scala]: Is the format string necessary here?

File: src/test/scala/no/gdl/bookapi/service/FeedServiceTest.scala [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+import no.gdl.bookapi.{BookApiProperties, TestData, TestEnvironment, UnitSuite}
+import org.mockito.Matchers._
+import org.mockito.Mockito._
+import scalikejdbc.DBSession
+
+class FeedServiceTest extends UnitSuite with TestEnvironment {
+
+  override val feedService = new FeedService
+
+  test("that calculateFeedUuids returns correct urls for one language and one level") {
```
Comment (runarfu) [inline src/test/scala/no/gdl/bookapi/service/FeedServiceTest.scala]: s/calculateFeedUuids/calculatedFeedUrls

File: src/main/resources/messages.txt [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+opds_root_title=Global Digital Library - Book Catalog
+
+new_entries_feed_title=New Entries
+new_entries_feed_description=The newest entries in the Global Digital Library
+
+level_feed_title=Level {0}
+level_feed_description=Books for selected level
+
+featured_feed_title=Editors picks
+featured_feed_description=Editors picks
```
[the remark this answers is not in the document]
Comment (kstigen) [reply src/main/resources/messages.txt]: I think it actually should be `Editors' picks`

File: src/main/scala/no/gdl/bookapi/service/FeedService.scala (hunk 5 above)
Comment (kstigen) [reply src/main/scala/no/gdl/bookapi/service/FeedService.scala]: Nope 👍

**Pull request merged**